cli/internal/core/engine_persistent_deps_test.go

```diff
@@ -0,0 +1,233 @@
+package core
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	testifyAssert "github.com/stretchr/testify/assert"
+	"github.com/vercel/turbo/cli/internal/fs"
+	"github.com/vercel/turbo/cli/internal/graph"
+	"github.com/vercel/turbo/cli/internal/util"
+
+	"github.com/pyr-sh/dag"
+)
+
+var WorkspaceGraphDefinition = map[string][]string{
+	"workspace-a": {"workspace-c"}, // a depends on c
+	"workspace-b": {"workspace-c"}, // b depends on c
+	"workspace-c": {},
+}
+
+func TestPrepare_PersistentDependencies_Topological(t *testing.T) {
+	completeGraph, workspaces := _buildCompleteGraph(WorkspaceGraphDefinition)
+
+	engine := NewEngine(&completeGraph.TopologicalGraph)
+
+	// "dev": dependsOn: ["^dev"] (where dev is persistent)
+	engine.AddTask(&Task{
+		Name:       "dev",
+		TopoDeps:   util.SetFromStrings([]string{"dev"}),
+		Deps:       make(util.Set), // empty, no non-caret task deps.
+		Persistent: true,
+	})
+
+	opts := &EngineBuildingOptions{
+		Packages:  workspaces,
+		TaskNames: []string{"dev"},
+	}
+	if err := engine.Prepare(opts); err != nil {
```

Comment:
[assert.NilError](https://pkg.go.dev/gotest.tools/v3/assert#NilError) (from gotest not testify 🙃 )is a nice helper method for checks like these

Reply:
updated! one thing that was kind of annoying was that you have to use like this:

```go
assert.Nil(t, engine.Prepare()) 
```
// this hides the real thing that you want to do

or
```go
err := engine.Prepare()
assert.NilError(t, err)
```

This pattern means `err` is declared for the rest of the block and I can't use the same pattern again. With:
```go
if err := engine.Prepare(); err != nil {}
```

`err` is only declared for this if block and the same pattern can be used again later.